examples/middleware/multiple_pipelines/src/main.rs

```diff
@@ -0,0 +1,270 @@
+//! Provides examples of using multiple middleware pipelines for different routes.
+//!
+//! We'll create an app with a few different routes that need
+//! different combinations of middleware.
+//!
+//! By default, we'll expect users to be logged in, using a session middleware
+//! to track client cookies.
+//! On the homepage, however, we don't want users logged in, so we'll override
+//! this path to not use any middleware.
+//! Our app also has some admin functionality, so we'll also override admin paths
+//! to require an additional admin session.
+//! Finally, our app exposes an JSON endpoint, which needs its own middleware.
+extern crate futures;
+extern crate gotham;
+#[macro_use]
+extern crate gotham_derive;
+
+#[macro_use]
+extern crate serde_derive;
+extern crate hyper;
+extern crate mime;
+
+use futures::future::{self, Either};
+use hyper::header::{HeaderMap, ACCEPT};
+use hyper::{Body, Response, StatusCode};
+
+use gotham::handler::HandlerFuture;
+use gotham::helpers::http::response::create_response;
+use gotham::middleware::session::NewSessionMiddleware;
+use gotham::middleware::Middleware;
+use gotham::pipeline::new_pipeline;
+use gotham::pipeline::set::{finalize_pipeline_set, new_pipeline_set};
+use gotham::pipeline::single::single_pipeline;
+use gotham::router::builder::*;
+use gotham::router::Router;
+use gotham::state::{FromState, State};
+
+/// A simple struct to represent our default session data.
+#[derive(Default, Serialize, Deserialize)]
+struct Session;
+
+/// We're going to define a separate struct for
+/// admin session data, to apply on the admin routes
+/// in addition to the default session data.
+#[derive(Default, Serialize, Deserialize)]
+struct AdminSession;
+
+/// As well as the session middlewares defined above,
+/// our example will provide an API route that uses
+/// its own middleware.
+#[derive(Clone, NewMiddleware)]
+pub struct ApiMiddleware;
+
+/// Our example API middleware will reject any requests that
+/// don't accept JSON as the response content type.
+impl Middleware for ApiMiddleware {
+    fn call<Chain>(self, state: State, chain: Chain) -> Box<HandlerFuture>
+    where
+        Chain: FnOnce(State) -> Box<HandlerFuture> + 'static,
+    {
+        let accepts = HeaderMap::borrow_from(&state)
+            .get(ACCEPT)
+            .map(|ct| ct.to_str().unwrap().to_string());
+
+        let f = match accepts {
+            None => Either::A(chain(state)),
```

Comment:
I think you could technically avoid the `Either` if you just return `Box::new(chain(state))` from this branch, and a Box directly in the other.

Not a big deal, but might be less noisy.

Reply:
Thanks - it was a while ago - but I prob got confused about mismatching future types.
Good to know this is the better solution.